### PR TITLE
Refactor/metabolite

### PR DESCRIPTION
**Pull request by akaviaLab**
* [X] description of feature/fix
Refactor metabolite.py and test_metabolite.py to be Python 3.6+ compliant including
removing future, typing, removing bugbears, documentation

### CODE REVIEW

File: src/cobra/core/metabolite.py [hunk 1]
```diff
@@ -40,7 +44,30 @@ class Metabolite(Species):
...
 
-    def __init__(self, id=None, formula=None, name="", charge=None, compartment=None):
+    # noinspection PyShadowingBuiltins
+    def __init__(
+        self,
+        id: str = None,
+        formula: str = None,
+        name: str = "",
+        charge: float = None,
+        compartment: str = None,
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]: Always make type annotations `Optional` when the argument can be `None`, please.

File: src/cobra/core/metabolite.py [hunk 2]
```diff
@@ -40,7 +44,30 @@ class Metabolite(Species):
...
+    ):
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]:
```suggestion
    ) -> None:
```

File: src/cobra/core/metabolite.py [hunk 3]
```diff
@@ -40,7 +44,30 @@ class Metabolite(Species):
...
+        formula : str
+            Chemical formula (e.g. H2O)
+        name : str
+            A human readable name.
+        charge : float
+           The charge number of the metabolite
+        compartment: str or None
+           Compartment of the metabolite.
+        """
         Species.__init__(self, id, name)
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]:
```suggestion
        super().__init__(id=id, name=name)
```

File: src/cobra/core/metabolite.py [hunk 4]
```diff
@@ -49,31 +76,48 @@ def __init__(self, id=None, formula=None, name="", charge=None, compartment=None
...
             the optlang constraint for this metabolite
         """
         if self.model is not None:
             return self.model.constraints[self.id]
 
     @property
-    def elements(self):
-        """Dictionary of elements as keys and their count in the metabolite
-        as integer. When set, the `formula` property is update accordingly"""
+    def elements(self) -> Union[None, Dict[Any, Union[int, float]]]:
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]:
```suggestion
    def elements(self) -> Optional[Dict[str, Union[int, float]]]:
```

Comment (Midnighter) [reply src/cobra/core/metabolite.py]: Or can it be something other than `str`?

File: src/cobra/core/metabolite.py [hunk 5]
```diff
@@ -251,30 +303,23 @@ def summary(self, solution=None, fva=None):
...
             <tr>
-                <td><strong>Metabolite identifier</strong></td><td>{id}</td>
+                <td><strong>Metabolite identifier</strong></td><td>{self.id}</td>
             </tr><tr>
-                <td><strong>Name</strong></td><td>{name}</td>
+                <td><strong>Name</strong></td><td>{format_long_string(self.name)}</td>
             </tr><tr>
                 <td><strong>Memory address</strong></td>
-                <td>{address}</td>
+                <td>{"0x0%x" % id(self)}</td>
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]:
```suggestion
                <td>{id(self):#x}</td>
```

**Review by Midnighter — CHANGES_REQUESTED** [review]
Overall, it looks good. I'm in doubt about the `Any` again. Would have to look more thoroughly at the code myself to make a call. Which is also an option. I can add fix ups afterwards if you prefer.

File: src/cobra/core/metabolite.py (hunk 4 above)
Comment (akaviaLab) [reply src/cobra/core/metabolite.py]: I think that metabolite is only str, but DictList (as defined) is more complicated. I'll go into detail why I used Any in dictlist in that PR.

File: src/cobra/core/metabolite.py [hunk 6]
```diff
@@ -40,40 +44,80 @@ class Metabolite(Species):
...
             the optlang constraint for this metabolite
         """
         if self.model is not None:
             return self.model.constraints[self.id]
 
     @property
-    def elements(self):
-        """Dictionary of elements as keys and their count in the metabolite
-        as integer. When set, the `formula` property is update accordingly"""
+    def elements(self) -> Union[None, Dict[str, Union[int, float]]]:
```
Comment (Midnighter) [inline src/cobra/core/metabolite.py]:
```suggestion
    def elements(self) -> Optional[Dict[str, Union[int, float]]]:
```

**Review by Midnighter — APPROVED** [review]
LGTM

**Pull request merged**